### PR TITLE
Perf: optimize scalar multiplication for 2-chains

### PR DESCRIPTION
**Pull request by yelhousni**
# Description

Similar idea as #1061 applied to 2-chains. But mulitplexers are costly in native arithmetic so the lookup idea is limited to only `varScalarMul` with `Lookup2`. 

 
## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

Current tests are passing.

# How has this been benchmarked?

e.g. BLS12-377 on BW6-761:

- G1 `varScalarMul` goes from 3,851 scs to **3,587 scs**.
- PLONK verifier circuit goes from 372,598 scs to **367,119 scs**.

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by yelhousni** [comment]
@ivokub the PR is good to go on my end.

**Review by ivokub — APPROVED** [review]
I'm not fully sure though in what cases we need to check GLV scalar decomposition `s1 + lambda * s2 == fr*s3 + s` and when not. Dunno if it would be safer if we use field emulation equality assertion to avoid any potential edge cases due to malicious hint?

But otherwise, it looks good. I think we can merge as is particularly that we have tests for all the edge cases.

File: std/algebra/native/sw_bls24315/g1.go [hunk 1]
```diff
@@ -190,77 +191,75 @@ func (P *G1Affine) varScalarMul(api frontend.API, Q G1Affine, s frontend.Variabl
...
-	api.AssertIsEqual(api.Add(s1, api.Mul(s2, cc.lambda)), api.Add(s, api.Mul(cc.fr, sd[2])))
-
-	// As the decomposed scalars are not fully reduced, then in addition of
-	// having the high bit set, an overflow bit may also be set. Thus, the total
-	// number of bits may be one more than the bitlength of λ.
-	nbits := cc.lambda.BitLen() + 1
+	// s1 + λ * s2 == s mod r
+	api.AssertIsEqual(
+		api.Add(s1, api.Mul(s2, cc.lambda)),
+		api.Add(s, api.Mul(cc.fr, sd[2])),
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/native/sw_bls24315/g1.go:204]: Also BLS24 code was a copy paste of BLS12 but things are a bit different. The eigenvalue lambda in BLS12 happen to have a bitsize equal to half the bitsize of the scalar field (because it happens to be x₀²-1 where x₀ is 64 bits) but for BLS24 it's not (it's x₀⁸ where x₀⁸ is 32 bits). So when we do the trick of adding lambda and lambda+1 to the sub-scalars we loose all the benefit of GLV because the loop would have the same size as the original scalar s and not half of it.

File: std/algebra/native/sw_bls12377/hints.go [hunk 2]
```diff
@@ -18,44 +20,71 @@ func init() {
...
+	}
+	cc := getInnerCurveConfig(scalarField)
+	sp := ecc.SplitScalar(inputs[0], cc.glvBasis)
+	outputs[0].Set(&(sp[0]))
+	outputs[1].Set(&(sp[1]))
+
+	return nil
+}
+
+func decomposeScalarG1(scalarField *big.Int, inputs []*big.Int, outputs []*big.Int) error {
```
[the remark this answers is not in the document]
Comment (yelhousni) [reply std/algebra/native/sw_bls12377/hints.go:38]: For `jointScalarMulUnsafe` we still need the old version that sets either of the top bits of the GLV sub-scalars. The loopkup-based new version is not efficient for `jointScalarMulUnsafe` as it would require multiplexers which are costly compared to native point arithmetic.

[the remark this answers is not in the document]
Comment (ivokub) [reply std/algebra/native/sw_bls12377/hints.go:38]: I see!

File: std/algebra/native/sw_bls24315/g1.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ivokub) [reply std/algebra/native/sw_bls24315/g1.go:204]: I see, it makes complete sense.
